proposals/NNNN-package-manager-product-definitions.md

```diff
@@ -110,41 +158,32 @@ Package(
 The definition of the `Product` type will be:
    
 ```
-public class Product {
-    public let name: String
-    public let targets: [String]
+public enum Product {
+    public class AbstractProduct {
+        public let name: String
+        public let targets: [String]
+        
+        public init(name: String, targets: [String])
+    }
     
-    public init(name: String, targets: [String]) {
-        self.name = name
-        self.targets = targets
+    public class Executable : AbstractProduct {
+        public init(name: String, targets: [String])
+    }
+    
+    public class Library : AbstractProduct {
+        public let type: LibraryType?
+    
+        public init(name: String, type: LibraryType? = nil, targets: [String])
     }
-}
-
-public class ExecutableProduct : Product {
-    // nothing else at this time
-}
     
-public class LibraryProduct : Product {
     public enum LibraryType {
```

Comment:
It used to be inside `Library`, but I thought `Library.LibraryType` seemed silly (and we can't name it `Type` since that's reserved).  I wouldn't mind pulling it back inside the class though.